pulp_ansible/app/models.py

```diff
@@ -266,6 +298,312 @@ class Meta:
         unique_together = ("pubkey_fingerprint", "signed_collection")
 
 
+class SigstoreSigningService(Content):
```

Comment:
This is most probably not a type of Content. Can you look into the signing services concept in Pulp?
I'm wondering how much of the vast amount of configuration can be handled on the shellscript level a signing service calls into. Also please sepearate the concept of signing from the concept of verifying.
Am i seeing this right that Sigstore creates x509 instead of gpg signatures? If that is correct, all the verification part should know of is whether a x509 cert is valid for some root cert.

I may be missing a point, but if think the signing service should be a x509-signing-service and using sigstore, openssl, a hardware cryptodevice or whatever cert authority is merely an implementation detail of the script called into.

Reply:
> Do you think the generic parts can be moved to pulpcore?

@ipanova The configuration part in SigstoreSigningService (model fields) could eventually be reused by other plugins as they are only relative to which Sigstore instances to connect to, which kind of verification to perform etc. The sigstore_sign and sigstore_verify methods could also be reused as AFAIK, the verification materials do not differ for different artifact types.

> I'm wondering how much of the vast amount of configuration can be handled on the shellscript level a signing service calls into.

@mdellweg The idea behind Sigstore signing services is to offer a native integration of Sigstore using the sigstore-python library directly into Pulp instead of calling the cosign command line from a shell script as it is the case for GPG signing as of now. As Sigstore signing and verification can get pretty complex and necessitates among other things to call the different Sigstore instance APIs and handle a larger number of exceptions than GPG, it is probably better to have some control over the process.

> Am i seeing this right that Sigstore creates x509 instead of gpg signatures? If that is correct, all the verification part should know of is whether a x509 cert is valid for some root cert.

Sigstore signing process generates two (plus one optional) outputs: a signature file and an X509 signing certificate (eventually a "Sigstore bundle" file used for offline verification). The certificate is short-lived (valid for 10min) and the verification process involves verifying the certificate signature by a certificate chain linking back to Sigstore CA (Fulcio).

> I may be missing a point, but if think the signing service should be a x509-signing-service and using sigstore, openssl, a hardware cryptodevice or whatever cert authority is merely an implementation detail of the script called into.

I am not quite sure what such an implementation would look like.
The CA used by Sigstore has [its own specification for issuing signing certificates](https://docs.sigstore.dev/fulcio/certificate-issuing-overview), and the client for verifying the certificate claims, which is not compatible with certificate signing with openssl or any other certificate authority as far as I can tell.